opalog/opalog.peg

```diff
@@ -1,64 +1,77 @@
 {
-// Command json parses JSON as defined by [1].
+package opalog
+
 //
 // BUGS: the escaped forward solidus (`\/`) is not currently handled for strings.
 //
-// TODO: check if JSON's numbers are a subset of Go's numbers, since we are
-//       assuming they are.  Currently representing all numbers as float64.
-// TODO: think about relaxing a Object to allow numeric keys, as they are common in practice.
-// [1]: http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf
-package opalog
-
-func toIfaceSlice(v interface{}) []interface{} {
-    if v == nil {
-        return nil
-    }
-    return v.([]interface{})
-}
 }
 
-Prog <- _ vals:Term+ EOF {
-    if vals == nil {
+Prog <- _ head:Term tail:( ws Term )*  EOF {
+    if head == nil {
         return make([]interface{}, 0), nil
     }
-    return vals.([]interface{}), nil
+    tailSlice := tail.([]interface{})
+    return append([]interface{}{head}, tailSlice...), nil
 }
 
-Term <- val:( Object / Array / Number / String / Bool / Null ) _ {
+Term <- val:( Composite / Scalar / Var ) {
     return val, nil
 }
 
-Object <- '{' _ vals:( String _ ':' _ Term ( ',' _ String _ ':' _ Term )* )? '}' {
-    valsSl := toIfaceSlice(vals)
-    if len(valsSl) == 0 {
-        return NewTerm([]*Term{}, OBJECT, c.text, "", c.pos.line, c.pos.col), nil
+Composite <- Object / Array
+
+Scalar <- Number / String / Bool / Null
+
+Object <- '{' _ head:((Scalar / Var) _ ':' _ Term)? tail:( _ ',' _ (Scalar / Var) _ ':' _ Term )* _ '}' {
+
+    set := NewKeyValueSet()
+
+    // Empty object.
+    if head == nil {
+        return NewTerm(set, OBJECT, c.text, "", c.pos.line, c.pos.col), nil
     }
-    restSl := toIfaceSlice(valsSl[5])
-    // Storing object arguments as a set of KeyValue pairs since we may not be
-    //   able to evaluate the keys (e.g. the key may be a variable)
-    res := NewKeyValueSet()
-    res.Add(NewKeyValue(valsSl[0].(*Term), valsSl[4].(*Term)))
-    for _, v := range restSl {
-        vSl := toIfaceSlice(v)
-        res.Add(NewKeyValue(vSl[2].(*Term), vSl[6].(*Term)))
+
+    // Non-empty object, first key/value pair.
+    headSlice := head.([]interface{})
+    set.Add(NewKeyValue(headSlice[0].(*Term), headSlice[len(headSlice) - 1].(*Term)))
+
```

Comment:
Would be helpful to have a comment that shows an example of what 'head' looks like.  I was confused as to why it wasn't headSlice[0] and headSlice[1].